linux/devdoc/threadpool_linux_requirements.md

```diff
@@ -0,0 +1,154 @@
+
+MOCKABLE_FUNCTION(, THREADPOOL_HANDLE, threadpool_create, EXECUTION_ENGINE_HANDLE, execution_engine);
+MOCKABLE_FUNCTION(, void, threadpool_destroy, THREADPOOL_HANDLE, threadpool);
+
+MOCKABLE_FUNCTION(, int, threadpool_open_async, THREADPOOL_HANDLE, threadpool, ON_THREADPOOL_OPEN_COMPLETE, on_open_complete, void*, on_open_complete_context);
+MOCKABLE_FUNCTION(, void, threadpool_close, THREADPOOL_HANDLE, threadpool);
+
+MOCKABLE_FUNCTION(, int, threadpool_schedule_work, THREADPOOL_HANDLE, threadpool, THREADPOOL_WORK_FUNCTION, work_function, void*, work_function_context);
+
+MOCKABLE_FUNCTION(, int, threadpool_timer_start, THREADPOOL_HANDLE, threadpool, uint32_t, start_delay_ms, uint32_t, timer_period_ms, THREADPOOL_WORK_FUNCTION, work_function, void*, work_function_context, TIMER_INSTANCE_HANDLE*, timer_handle);
+
+MOCKABLE_FUNCTION(, int, threadpool_timer_restart, TIMER_INSTANCE_HANDLE, timer, uint32_t, start_delay_ms, uint32_t, timer_period_ms);
+
+MOCKABLE_FUNCTION(, void, threadpool_timer_cancel, TIMER_INSTANCE_HANDLE, timer);
+
+MOCKABLE_FUNCTION(, void, threadpool_timer_destroy, TIMER_INSTANCE_HANDLE, timer);
+
+```
+###  threadpool_create
+
+```C
+MOCKABLE_FUNCTION(, THREADPOOL_HANDLE, threadpool_create, EXECUTION_ENGINE_HANDLE, execution_engine);
+```
+
+`threadpool_create` creates a new threadpool.
+
+`threadpool_create` shall allocate memory for a threadpool object and on success return a non-NULL handle to it. 
+
+If `execution_engine` is `NULL`, `threadpool_create` shall fail and return `NULL`. 
+
+`threadpool_create` shall initialize thread count paramters by calling `execution_engine_linux_get_parameters` with paramter `execution_engine`. 
+
+`threadpool_create` shall allocate memory for a list of thread objects and on success return a non-NULL handle to it.
+
+`threadpool_create` shall allocate memory for a list of tasks and on success return a non-NULL handle to it.
+
+`threadpool_create` shall initialize every thread item in the threads array with `task_func` and `task_param` set to `NULL` and `task_state` set to `TASK_NOT_USED`.
+
+`threadpool_create` shall create a shared semaphore between threads with initialized value zero.
+
+`threadpool_create` shall initilize the `state` to `THREADPOOL_STATE_NOT_OPEN`.
+
+`insert_idx`  and `consume_idx` shall be intialzied to -1 to make the first increment start at zero.
+
+`threadpool_create` shall destory the semaphore after initialization.
+
+If any error occurs, `threadpool_create` shall fail and return `NULL`.
+
+### threadpool_destroy
+
+```C
+MOCKABLE_FUNCTION(, void, threadpool_destroy, THREADPOOL_HANDLE, threadpool);
+```
+
+`threadpool_destroy` frees the resouces associated with `threadpool`.
+
+`threadpool_destroy` shall free the resources associated with the threadpool handle. 
+
+If `threadpool` is `NULL`, `threadpool_destroy` shall return.
+
+`threadpool_destroy` shall close the threadpool by calling `internal_close`. 
```

Comment:
>`threadpool_destroy` shall close the threadpool by calling `internal_close`. [](http://example.com/codeflow?start=0&length=76)

Don't really need to detail this one, it would be better to say

`threadpool_destroy` shall implicit close if `threadpool` state is set to `THREADPOOL_STATE_OPEN`